### PR TITLE
Fix UPF binary API and add dump calls

### PR DESCRIPTION
**Pull request by ivan4th**
Fix string handling in the API. They were being confused with vectors.
Fix deletion of applications. The hash key was being removed too
early.
Add upf_applications_dump and upf_application_l7_rule_dump calls
to make it possible to do proper sync with the controller.

### CODE REVIEW

File: src/plugins/upf/upf_adf.c [hunk 1]
```diff
@@ -522,10 +521,12 @@ vnet_upf_app_add_del (u8 * name, u32 flags, u8 add)
       }));
       /* *INDENT-ON* */
 
+      hash_unset_mem (sm->upf_app_by_name, app->name);
       upf_adf_remove (app->db_index);
       vec_free (app->name);
       hash_free (app->rules_by_id);
       pool_free (app->rules);
+      clib_memset (app, 0, sizeof (*app));
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_adf.c:529]: is that memset truly needed?

File: src/plugins/upf/upf_api.c [hunk 2]
```diff
@@ -96,9 +99,11 @@ vl_api_upf_app_add_del_t_handler (vl_api_upf_app_add_del_t * mp)
   vl_api_upf_app_add_del_reply_t *rmp = NULL;
   upf_main_t *sm = &upf_main;
   int rv = 0;
+  u8 *name = format(0, "%s", mp->name);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_api.c:102]: there is a simpler vec function for this,

```
vec_add (name, mp->name, strlen (mp->name));
```

Assuming that mp->name is a zero terminated C string.

Comment (ivan4th) [reply src/plugins/upf/upf_api.c:102]: Same, I've basically copied that from other plugins. I'm not very well-versed in VPP's clib atm, will look for a simpler way to do it

File: src/plugins/upf/upf_adf.c (hunk 1 above)
Comment (ivan4th) [reply src/plugins/upf/upf_adf.c:529]: That's what other plugins are doing. I'd say it's not strictly needed but zeroing out objects before returning them to the pool may make attackers' life harder in some scenarios. This is probably a bit too far-fetched (if an attacker has access to this API he basically owns the VPP instance anyway), so if you think that's redundant, I can remove that.

File: src/plugins/upf/upf_api.c (hunk 2 above)
Comment (ivan4th) [reply src/plugins/upf/upf_api.c:102]: Actually, I can see `vec_validate_init_c_string()` macro, which is used just once, and some code just uses `vac_validate()`, neither being more concise than this `format()` call, performance issues from calling `format()` not being relevant to these relatively rare API calls, so I'd appreciate any pointers concerning what function should I use. Re this `format()` trick, it seems to be used quite a bit in the VPP source:
```
git grep 'format (0, "%s"' | wc -l
      87
```

File: src/plugins/upf/upf_api.c [hunk 3]
```diff
@@ -136,11 +147,108 @@ static void vl_api_upf_app_flow_timeout_set_t_handler
...
+
+  mp = vl_msg_api_alloc (sizeof (*mp));
+  clib_memset (mp, 0, sizeof (*mp));
+
+  mp->_vl_msg_id = htons (VL_API_UPF_APPLICATIONS_DETAILS
+			  + sm->msg_id_base);
+  mp->context = context;
+
+  name_len = clib_min(vec_len(app_name) + 1, ARRAY_LEN(mp->name));
+  memcpy (mp->name, app_name, name_len-1);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_api.c:171]: you need to alloc mp->name

this should also work:
```
vec_validate_init_c_string(mp->name, app_name, vec_len(app_name));
```

File: src/plugins/upf/upf_api.c (hunk 2 above)
Comment (ivan4th) [reply src/plugins/upf/upf_api.c:102]: ... ok, I see, will fix that

File: src/plugins/upf/upf_api.c (hunk 3 above)
Comment (ivan4th) [reply src/plugins/upf/upf_api.c:171]: Well, `mp->name` is not a vector, it's just an array in the struct, as defined here

```
define upf_applications_details {
  u32 context;

  u8 name[64];
  u32 flags;
};
```

Comment (ivan4th) [reply src/plugins/upf/upf_api.c:171]: Here's similar piece of code in another plugin:
https://github.com/travelping/vpp/blob/feature/1908/upf/src/plugins/memif/memif_api.c#L329-L330

Comment (RoadRunnr) [reply src/plugins/upf/upf_api.c:171]: ah, right, didn't see that

File: src/plugins/upf/upf_api.c (hunk 2 above)
Comment (RoadRunnr) [reply src/plugins/upf/upf_api.c:102]: yeah, seems that lots of people just copy from each other. Leave it in if you want. The overhead of the format string parsing is probably not worth it.

File: src/plugins/upf/upf_adf.c (hunk 1 above)
Comment (RoadRunnr) [reply src/plugins/upf/upf_adf.c:529]: I've use something similar to catch bugs in the past. So its not totally redundant. I find it much more important to make sure we clear an entry when we get it from the pool.

**Pull request merged**